itkwidgets/integrations/__init__.py

```diff
@@ -1,60 +1,132 @@
+        to_skip = { d: sizes[d] <= max_size / 2 for d in previous.keys() }
+        scale_factor = {}
+        for dim in previous.keys():
+            if to_skip[dim]:
+                scale_factor[dim] = previous[dim]
+                continue
+            scale_factor[dim] = 2 * previous[dim]
+
+            sizes[dim] = int(sizes[dim] / 2)
+        previous = scale_factor
+        scale_factors.append(scale_factor)
+
+    return scale_factors
+
+from IPython.core.debugger import set_trace
 
 def _get_viewer_image(image):
-    if HAVE_ITK:
+    min_length = 64
+    multiscale = None
+    if isinstance(image, MultiscaleSpatialImage):
+        multiscale = image
+    # Todo: support for itkwasm.Image
+    if multiscale is None and HAVE_ITK:
         import itk
         if isinstance(image, itk.Image):
-            return itk_image_to_wasm_image(image)
-    if HAVE_VTK:
+            dimension = image.GetImageDimension()
+            size = np.array(itk.size(image))
+            scale_factors = []
+            dims = ('x', 'y', 'z')
+            previous = {'x': 1, 'y': 1, 'z': 1}
+            while (size > min_length).any():
+                to_skip = size <= size.max() / 2
+                scale_factor = {}
+                for dim in range(dimension):
+                    if to_skip[dim]:
+                        scale_factor[dims[dim]] = previous[dims[dim]]
+                        continue
+                    scale_factor[dims[dim]] = 2 * previous[dims[dim]]
+                    size[dim] = int(size[dim] / 2)
+                previous = scale_factor
+                scale_factors.append(scale_factor)
+
+            multiscale = itk_image_to_multiscale(image, scale_factors=scale_factors)
+    if multiscale is None and HAVE_VTK:
         import vtk
         if isinstance(image, vtk.vtkImageData):
-            return vtk_image_to_ndarray(image)
-    if HAVE_DASK:
-        import dask
+            spatial_image = vtk_image_to_spatial_image(image)
+            scale_factors = _spatial_image_scale_factors(spatial_image, min_length)
+            multiscale = to_multiscale(spatial_image, scale_factors)
+    if multiscale is None:
         if isinstance(image, dask.array.core.Array):
-            return dask_array_to_ndarray(image)
-    if HAVE_TORCH:
+            spatial_image = to_spatial_image(image)
+            scale_factors = _spatial_image_scale_factors(spatial_image, min_length)
+            multiscale = to_multiscale(spatial_image, scale_factors)
+    if multiscale is None:
+        if isinstance(image, zarr.Array):
+            spatial_image = to_spatial_image(image)
+            scale_factors = _spatial_image_scale_factors(spatial_image, min_length)
+            multiscale = to_multiscale(spatial_image, scale_factors)
+    if multiscale is None:
+        # NGFF Zarr
+        if isinstance(image, zarr.Group) and 'multiscales' in image.attrs:
+            return image.store
+    if multiscale is None and HAVE_TORCH:
```

Comment:
It is worth removing the `is multiscale` checks with a `func makeMultiscale(image)` ?
```
def _get_viewer_image(image):

Reply:
Good idea -- yes, that is a bit cleaner.